common/fixation_entry.go

```diff
@@ -0,0 +1,280 @@
+		}
+	case types.DO_NOTHING:
+	}
+
+	return nil
+}
+
+// GetStoreKey returns the Fixation store's store key
+func (fs *FixationStore) GetStoreKey() sdk.StoreKey {
+	return fs.storeKey
+}
+
+// GetCdc returns the Fixation store's codec
+func (fs *FixationStore) GetCdc() codec.BinaryCodec {
+	return fs.cdc
+}
+
+// Getprefix returns the Fixation store's fixation key
+func (fs *FixationStore) GetPrefix() string {
+	return fs.prefix
+}
+
+// getUnmarshaledEntryForBlock gets an entry by block. Block doesn't have to be precise, it gets the closest entry version
+func (fs *FixationStore) getUnmarshaledEntryForBlock(ctx sdk.Context, index string, block uint64, refAction types.ReferenceAction) (*types.Entry, error) {
+	// get the relevant store using index
+	store := prefix.NewStore(ctx.KVStore(fs.storeKey), types.KeyPrefix(fs.createStoreKey(index)))
+
+	// init a reverse iterator
+	iterator := sdk.KVStoreReversePrefixIterator(store, []byte{})
+	defer iterator.Close()
+
+	// iterate over entries
+	for ; iterator.Valid(); iterator.Next() {
+		// unmarshal the entry
+		var entry types.Entry
+		fs.cdc.MustUnmarshal(iterator.Value(), &entry)
+
+		// if the entry's block is smaller than or equal to the requested block, unmarshal the entry's data
+		// since the user might not know the precise block that the entry was created on, we get the closest one
+		// we get from the past since this was the entry that was valid in the requested block (assume there's an
+		// entry in block 100 and block 200 and the user asks for the version of block 199. Since the block 200's
+		// didn't exist yet, we get the entry from block 100)
+		if entry.GetBlock() <= block {
+			err := fs.handleRefAction(ctx, &entry, refAction)
+			if err != nil {
+				return nil, err
+			}
+			// get the relevant byte
+			byteKey := types.KeyPrefix(createEntryKey(entry.GetBlock()))
+
+			// marshal the entry
+			marshaledEntry := fs.cdc.MustMarshal(&entry)
+
+			// set the entry
+			store.Set(byteKey, marshaledEntry)
+			return &entry, nil
+		}
+	}
+
+	return nil, types.ErrEntryNotFound
+}
+
+// GetEntryForBlock gets an entry by block. The entry is returned by reference on the entryData argument. Block doesn't have to be precise, it gets the closest entry version
+func (fs *FixationStore) GetEntry(ctx sdk.Context, index string, block uint64, entryData codec.ProtoMarshaler, refAction types.ReferenceAction) error {
+	// get the unmarshaled entry for block
+	entry, err := fs.getUnmarshaledEntryForBlock(ctx, index, block, refAction)
+	if err != nil {
+		return utils.LavaError(ctx, ctx.Logger(), "GetEntry_cant_get_entry", map[string]string{"err": err.Error()}, "can't get entry")
+	}
+
+	// unmarshal the entry's data
+	err = fs.cdc.Unmarshal(entry.GetData(), entryData)
+	if err != nil {
```

Comment:
(See also comment in `getUnmarshaledEntryForBlock()`)

If `refAction` is to increase/decrease the count, and an error occurred here, then the refcount increase/decrease must be undone. 

It's better API to split into GetEntry (find+inc), FindEntry (find, no inc), and PutEntry (find, dec).
Also, it's confusing that a callers call GetEntry() to drop (a reference to) it.